### PR TITLE
Revamp Assets

### LINKED ISSUE #147: Resource Subsystem
Add system to load and cache resources, including zip safe.

### PR DESCRIPTION
**Pull request by astronouth7303**
Totally revamp how PPB handles assets.

Depends on #313 because life was just easier that way.

Part of #147

### CODE REVIEW

File: ppb/systems/__init__.py [hunk 1]
```diff
@@ -99,8 +101,9 @@ def prepare_rectangle(self, resource, game_object, camera):
 
     def register(self, resource_path, name=None):
```
Comment (pathunstrom) [inline ppb/systems/__init__.py]: I feel like we probably want to rename the parameter here? Since it's not resource_path anymore?

Comment (astronouth7303) [reply ppb/systems/__init__.py]: Oops! This code actually got totally nixed. I forgot to mark this PR as WIP.

File: tests/test_sprites.py [hunk 2]
```diff
@@ -326,18 +326,3 @@ def test_rotatable_base_sprite():
 
     test_sprite.rotate(1)
     assert test_sprite.rotation == 1
-
-
-def test_sprite_in_main():
```
Comment (pathunstrom) [inline tests/test_sprites.py]: Is there a solid way to keep a test around this? The use case of "no actual __file__" is still a possibility, we just resolve it differently.

File: ppb/systems/__init__.py [hunk 3]
```diff
@@ -78,14 +98,14 @@ def render_background(self, scene):
...
-        image_name = game_object.__image__()
-        if image_name is flags.DoNotRender:
+        image = game_object.__image__()
+        if image is flags.DoNotRender:
             return None
-        image_name = str(image_name)
-        if image_name not in self.resources:
-            self.register_renderable(game_object)
+        if isinstance(image, str):
+            logger.warn(f"Using string resources is deprecated, use ppb.Image instead. Got {image!r}")
```
Comment (pathunstrom) [inline ppb/systems/__init__.py]: 👍

File: ppb/systems/__init__.py [hunk 4]
```diff
@@ -22,6 +26,22 @@ def __exit__(self, exc_type, exc_val, exc_tb):
 
 
 from ppb.systems.pg import EventPoller as PygameEventPoller  # To not break old imports.
+from ppb.assets import Asset
+import io
+
+
+class Image(Asset):
```
Comment (pathunstrom) [inline ppb/systems/__init__.py]: Dunno on adding new things to `systems.__init__`. If we're making changes to the renderer and adding new things they should move to its own file.

File: ppb/systems/__init__.py (hunk 3 above)
Comment (astronouth7303) [reply ppb/systems/__init__.py]: Only thing about this is that it's extremely verbose (every frame).

File: ppb/systems/__init__.py (hunk 4 above)
Comment (astronouth7303) [reply ppb/systems/__init__.py]: Yeah, I was wondering if we wanted to move the renderer to its own module.

Comment (pathunstrom) [reply ppb/systems/__init__.py]: It's one of the places we deal with pygame, so probably needs to go to `systems.pg`. (There's even argument for making pg.py into a subpackage of its own, but uh. . . not right now.)

File: ppb/systems/__init__.py (hunk 3 above)
Comment (pathunstrom) [reply ppb/systems/__init__.py]: Could throw in a tracking variable and only throw it once? (Or once per object?) Would `lru_cache` wrapped around a print statement make it run once per object passed?

**Comment by astronouth7303** [comment]
This still needs tests.

File: ppb/systems/__init__.py (hunk 3 above)
Comment (astronouth7303) [reply ppb/systems/__init__.py]: Yeah, all of those things. It's just annoying, though?

**Comment by astronouth7303** [comment]
@pathunstrom I think this is ready for review & merge.

**Pull request merged**